File: WindowsApplication1/Utilities/ApplicationSettings.vb
﻿Imports System.Configuration

Public Class ApplicationSettings

    Public Enum Settings
        Version = 1
        DefaultWatchArgs = 2
        VLCPath = 3
        MPCPath = 4
        LiveStreamerPath = 5
        ServiioPath = 6
        ShowScores = 7
        RefreshIntervalInMin = 8
        ShowAdvancedWatchPanel = 9
    End Enum

    Public Shared Function Read(Of T)(key As Settings, Optional defaultReturnValue As Object = Nothing) As T
        Try
            Dim configFile = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None)
            Dim settings = configFile.AppSettings.Settings
            Dim result As Object = settings(key.ToString()).Value
            If String.IsNullOrEmpty(result) OrElse IsNothing(result) Then
                Return defaultReturnValue
            End If
            Return CType(result, T)
        Catch e As ConfigurationErrorsException
            Console.WriteLine("Error reading app setting: " & key)
            Return defaultReturnValue
        End Try
    End Function

    Public Shared Sub SetValue(key As Settings, value As String)
        Try
            Dim configFile = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None)
            Dim settings = configFile.AppSettings.Settings
            If IsNothing(settings(key.ToString())) Then
                settings.Add(key.ToString(), value)
            Else
                settings(key.ToString()).Value = value
            End If

            If value.Length > 200 Then
                value = "[Value too large for display]"
            End If
            Console.WriteLine("Status: Setting updated for """ & key.ToString() & """ to """ & value & """")
            configFile.Save(ConfigurationSaveMode.Modified)
            ConfigurationManager.RefreshSection(configFile.AppSettings.SectionInformation.Name)
        Catch e As ConfigurationErrorsException
            Console.WriteLine("Error writing app settings")
        End Try
    End Sub

End Class

File: NHLGames.WPF/Utilities/ApplicationSettings.cs
using System;
using System.Configuration;

namespace NHLGames.WPF.Utilities
{
    public class ApplicationSettings
    {
        public static T Read<T>(SettingsEnum key, object defaultReturnValue = null)
        {
            try
            {
                var configFile = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None);
                dynamic result = configFile.AppSettings.Settings[key.ToString()].Value;
                if (string.IsNullOrEmpty(result))
                {
                    return (T)defaultReturnValue;
                }

                if (result is T variable)
                {
                    return variable;
                }


                if (bool.TryParse(result, out bool boolResult))
                {
                    return (T) Convert.ChangeType(boolResult, typeof(T));
                }

                try
                {
                    return Serialization.DeserializeObject<T>(result);
                }
                catch (Exception)
                {
                    Console.WriteLine(Properties.Resources.errorDeserialize, key.ToString(), typeof(T));
                    return (T) defaultReturnValue;
                }

            }
            catch (ConfigurationErrorsException)
            {
                Console.WriteLine(Properties.Resources.errorReadingSettings, key);
                return (T) defaultReturnValue;
            }
        }


        public static void SetValue(SettingsEnum key, string value)
        {
            try
            {
                var configFile = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None);
                var settings = configFile.AppSettings.Settings;
                if (settings[key.ToString()] == null)
                {
                    settings.Add(key.ToString(), value);
                }
                else
                {
                    settings[key.ToString()].Value = value;
                }

                if (value != null && value.Length > 200)
                {
                    value = Properties.Resources.msgValueTooLarge;
                }
                if (key != SettingsEnum.DefaultWatchArgs)
                {
                    Console.WriteLine(Properties.Resources.msgSettingUpdated, key.ToString(), value);
                }
                configFile.Save(ConfigurationSaveMode.Modified);
                ConfigurationManager.RefreshSection(configFile.AppSettings.SectionInformation.Name);
            }
            catch (ConfigurationErrorsException)
            {
                Console.WriteLine(Properties.Resources.errorWritingSettings);
            }
        }

    }
}
